src/main/scala/mesosphere/marathon/state/Volume.scala

```diff
@@ -136,7 +136,7 @@ case class DiskSource(
       else
         bld.setPath(Source.Path.newBuilder().setRoot(p))
       id.foreach(bld.setId)
-      bld.setMetadata(metadata.toMesosLabels)
+      metadata.foreach(metadata => bld.setMetadata(metadata.toMesosLabels))
```

Comment:
This doesn't safely improve the situation IMO. If metadata is passed as `Some(Map.empty)`, it will still lead to the incompatible behavior. That might not be the case now, but it could be introduced as a regression. Using `metadata: Map[String, String]` and checking for `.nonEmpty` would prevent that.

Otherwise we should prevent that `DiskSource` ever provides a `Some(Map.empty)`. From what I can see, `DiskSource` doesn't need to be a case class. Private constructor and an `apply` that makes sure the metadata is populated correctly?

Reply:
@meichstedt, the thing is that if `metadata` is empty in a resource offer, it should be empty in an offer operation; if `metadata` is absent in a resource offer, it should be empty in an offer operation too. That's why its type has been changed to `Option[Map[String, String]]`. Does it address your concern? I am not sure if we refer to the same thing here.

Reply:
@meichstedt I don't follow. Marathon should reliably reproduce what Mesos encodes in the originally offered resource.disk.source.metadata. `metadata` is of type `Labels`, which is a `message`, not `repeated` or `map`.

1. no `metadata` protobuf <--> `None`
2. an empty `metadata` protobuf <--> `Some(Map.empty)`
3. non-empty `metadata` protobuf <--> `Some(Map[String,String] { ... entries ... })`

Reply:
@jdef I agree – when reviewing, I was under the impression that there's a chance Marathon mutates those values. It doesn't so that's fine. @ichernetsky created a JIRA a while back related to this ser/deser being a problem in general, since we potentially lack information afterwards.